[IMP] website, *: improve the "edit in backend" button

Guillaume-gdi opened the pull request:
*: base, mrp_account, product, mail, website_blog, website_event,
   website_forum

This commit improves the button that allows to go to the backend view of
an object when you are on its corresponding page on the website. The
button is more visible and the user can see which object he is going to
edit. Note that this commit also:

- Changes the access key to translate a website page to ALT + T.

- Adds a new access key to edit an object in backend with ALT + E.

- Removes the possibility to duplicate a blog post (but this feature
will be reintroduced later for all models, generically) **.

**: Note that the duplication of blog posts actually had a mistake:

The controller to create a new blog post has been added with [1] where
it has been decided to not be a follower of the blog posts at their
creation. A new controller has been added by [2] to be able to duplicate
a blog post, to be consistent with [1], here also the user does not
become a follower of the new blog post (the copy). So far, so good.
Finally, [3] has changed the blog post creation controller so that the
user who creates the blog post is a follower of the new blog post.
Unfortunately the same change was not made for the duplicate controller,
which is a mistake. There is no reason to be a follower of the newly
created blog posts when you go through the add blog post controller but
not when you go through the duplication controller. The behaviors should
be consistent and there is no reason for there to be a difference.

[1]: https://github.com/odoo/odoo/commit/4c3b516a7b988d758a67ff19242e8ed0837d756c
[2]: https://github.com/odoo/odoo/commit/fe40538aff2b65f7719840c8e2d6e51e858f560f
[3]: https://github.com/odoo/odoo/commit/4bf9dc4078a5ca413539a6fd16400fd5aad46907

task-2889929

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/97353).

rdeodoo commented on addons/website_event/static/src/js/systray_items/edit.js:
```diff
@@ -0,0 +1,16 @@
+     */
+    getEditActionName() {
+        if (this.websiteService.currentWebsite.metadata.mainObject.model === 'event.event') {
+            return this.env._t('Edit event');
```
Why are we doing this in JS?
This should be done in a generic way in python/views, `ir.model.name` is exactly what we want/need: the model's name which is translatable.
And it will work everywhere, even custos.

Guillaume-gdi replied:
Indeed it would be an elegant way to do it but I think that doing it will cause a translation problem. I agree that it will work well for most languages but not for all... For example, in Japanese:
`Edit` => `編集`
`Product` => `製品情報`
but `Edit product` should be translated as `製品を編集する` which is different from `編集` + ' ' +  `製品情報`.
I don't know if my example is correct and I don't know if this problem is important enough to justify not doing what you propose above... just to let you know

qsm-odoo replied:
Yep, I had the same concern with this suggestion of using `ir.model.name` since the beginning. Maybe instead of "Edit XXX", we could show "XXX" next to an edit-link icon like fa-pencil and that would be clear enough :shrug: I don't know, could use @pierre-pde opinion here.

rdeodoo replied:
I don't understand the japanese issue, the translatable string will be "Edit %s", which is no different from every single translatable string in Odoo, so.. Nothing too critical IMO.

But in any case I wouldn't do a "per module" override, I would make it generic, QSM suggestion is fine to me too :+1: 

Personal preference:
1. Do it generic, send "Edit %s" in translation, where %s is the main_object
2. Do it generic, "<fa-pencil> <model_name>"
3. The current way, per module override, but we won't do it anyway, generic is always better :+1:

qsm-odoo replied:
> the translatable string will be "Edit %s", which is no different from every single translatable string in Odoo

Yep, we can definitely fallback to "Edit %s" to be generic and handle custos but I think that a perfect translation for the main buttons of the UI is important IMO (so for "Edit product", I would definitely prefer having it translated this way rather than "Edit %s" if we have to have the "Edit" word). 

If we want to ship this before PDE is back, let's go for the solution I suggested: an icon + the model name, we can switch to (1) or (3) later if needed.

pierre-pde replied:
I prefer solution 1 proposed by RDE. it's fine if a translator has to translate the word 'edit' multiple times :D 
The tool for translation helps them anyway when the recognize something.

pierre-pde replied:
> I prefer solution 1 proposed by RDE. it's fine if a translator has to translate the word 'edit' multiple times :D The tool for translation helps them anyway when the recognize something.

@Guillaume-gdi FYI

qsm-odoo replied:
> I prefer solution 1 proposed by RDE. it's fine if a translator has to translate the word 'edit' multiple times :D

@pierre-pde You are confusing :grimacing: Solution 1 of RDE would not make the translator translate "edit" multiple times :shrug: Only the third solution kinda would but not really (would be translating sentences containing the "edit" word but would not necessarily translated to the same thing).

@Guillaume-gdi But I confirm that, for now, we go with solution 1. We may decide to override that default behavior per app later (but likely not, only I seems to think it's important so not worth it).

rdeodoo commented on addons/website/static/src/systray_items/edit.js:
```diff
@@ -0,0 +1,114 @@
+
+    getEditActionName() {
+        // Fallback if the function is not overridden.
+        return this.env._t('Edit in backend');
```
We don't want to reintroduce this term "In backend" when we are already in backend and ou clients don't know what is a "backend"

rdeodoo reviewed (changes requested):
Just a very, very quick one quickly flying over the files before the week end.
Just checked the wording and the implementation idea for the "Edit X" model

Guillaume-gdi commented:
Ok I implemented the solution. Note that I still patched two model names for this edit button: `Product Template` => `Product` and `Appointment Type` => `Appointment`. I hesitated to do it so that `Job Position` becomes `Job Offer` because it is this name that is in the "Add New Content" menu. But I changed the term in this Add New Content menu instead to make it consistent. I also added icons so that each element in the dropdown has an icon (I thought it would look bit ugly if only one element had an icon). Let me know if you want me to change anything :+1: .
@rdeodoo @qsm-odoo

rdeodoo commented on addons/website/views/website_templates.xml:
```diff
@@ -53,6 +53,7 @@
             'data-is-published': 'website_published' in main_object.fields_get() and main_object.website_published,
             'data-can-publish': 'can_publish' in main_object.fields_get() and main_object.can_publish,
             'data-editable-in-backend': editable_in_backend,
+            'data-object-name': main_object._description,
```
```suggestion
            'data-main-object-name': main_object._description if main_object else None,
```
(and move it next to the `data-main-object` line above)

Maybe we should just add this to the main-object, which should become `ir.ui.view/53/View`.
That would require a bit more change in Odoo as some code would need to be adapted.

So instead of
```html
<html data-main-object="ir.ui.view(3552,)" data-object-name="View"/>
<html data-main-object="product.template(9,)" data-object-name="Product Template"/>
```
=>
```html
<html data-main-object="product.template/9/Product Template"/>
```

It feels weird to have 2 properties out of 3 in a single attribute, I'd either have all 3 in separate attributes or all 3 in a single one?
Anyway, this might just be done later if we don't want to slow down this one I guess, at least just process the initial review of this message

rdeodoo commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -80,6 +80,18 @@ def jobs_add(self, **kwargs):
+            new_name = _("%s (copy %s)") % (job.name, str(counter))
+            counter += 1
+        job = job.copy(default={'name': new_name})
+        return request.redirect(f'@{job.website_url}?&enable_editor=1')
```
```suggestion
        return request.redirect(f'/@{job.website_url}?&enable_editor=1')
```

rdeodoo commented on addons/website/static/src/systray_items/edit.js:
```diff
@@ -0,0 +1,117 @@
+/** @odoo-module **/
```
This is a merge of edit_website.js and edit_in_backend.js + I guess a few changes you made, this is making the review really more difficult than it should IMO.
Could you please make a first commit where you merge those 2 files, it can be a temp commit during the review or even a [MOV] commit that we will keep (and it will "save" the git history of one of the 2 files)

Probably same for the xml file but less important as quite small
Same for the website_blog file

rdeodoo commented on addons/website_blog/static/src/js/systray_items/edit.js:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * @override
+     */
+    getElements() {
+        const elements = this._super();
+        const { metadata: { mainObject } } = this.websiteService.currentWebsite;
+        if (mainObject.model === 'blog.post') {
+            return [...elements, {
+                title: this.env._t('Duplicate'),
+                callback: () => this.duplicate(),
+                icon: 'fa-clone',
+            }];
+        }
+        return elements;
+    },
+    async duplicate() {
+        if (this.websiteService.currentWebsite.metadata.mainObject.model === 'blog.post') {
+            const { metadata: { mainObject } } = this.websiteService.currentWebsite;
+            wUtils.sendRequest('/blog/post_duplicate', { blog_post_id: mainObject.id });
+        } else {
+            this._super();
+        }
+    },
```
Red flag, this whole part is duplicate code in hr job
And it will be duplicated everytime we want to make a model able to be duplicated.

Move this in the `EditSystray` and make it used only by the hr job and blog, so it will be just 2-3 lines in case a model want to add the duplicate capability

Also, as the code will be generic in editsystay, make it generic in python too.
Eg, create a controller in website `/website/duplicate` like we have for `/website/publish`, I'll let you check.
You could then move the logic about finding a new unique name inside it (even if there is no uniq constraint on the model, it will still be nice to not have the exact same name on duplicate) + redirect to the website_url if the model has that property.
(the controller in blog will just be removed then)

You just need to properly implement something for the `.with_context(mail_create_nosubscribe=True)` part but maybe you can do that in JS instead in the small overide you will have here (to allow the duplicate capability, maybe also a getCopyContext method)

rdeodoo commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -80,6 +80,18 @@ def jobs_add(self, **kwargs):
         })
         return request.redirect(request.env["website"].get_client_action_url(f"/jobs/detail/{slug(job)}", True))
 
+    @http.route('/jobs/duplicate', type='http', auth="user", methods=['POST'], website=True)
+    def jobs_duplicate(self, job_id, **post):
+        job = request.env['hr.job'].browse(int(job_id))
```
Maybe the route could have been changed to include a model converter, and the loop can probably be improved but didn't check further, I think this should be removed and made generic (see other comment on JS part)

Guillaume-gdi commented:
@rdeodoo When you have time you can check this PR and [the enterprise PR](https://github.com/odoo/enterprise/pull/30075) that comes with it :)

Guillaume-gdi commented:
@rdeodoo Note that I can add the `me-X` class on the Dropdown to avoid the problem of positioning the dropdown items. But then the dropdown items list will be slightly far from the right edge... maybe it can act as a temporary (ugly) fix.

rdeodoo commented:
The specs changed again today following AL/FP request, we need to discuss this one but:
It shouldn't be a dropdown (good, I didn't like the conditional UX) but should now be a second button next to Edit with an icon and just the model name (it will flicker from page to page and could be extra long in some lang but..).
I guess it will solve the current problem of people not finding the "Edit Product" in the dropdown (I am not sure it is a real problem tho).

pierre-pde commented:
@Guillaume-gdi , are you aware of the change that Romain described? don't hesitate to call me if you have any question.
The icon should before 'Product' (or other object) should probably be fa-cog

Guillaume-gdi commented:
Hello @pierre-pde. I am only aware of the above message. I will have a meeting with @rdeodoo today to discuss the new specs.

pierre-pde commented:
@Guillaume-gdi  & @rdeodoo , from what I've undertsood, here is the new spec;
https://bit.ly/3Tpi1rj

Guillaume-gdi commented:
I have adapted this PR to the new specs. :+1: 
cc @rdeodoo @qsm-odoo @pierre-pde

qsm-odoo commented on addons/website/views/website_templates.xml:
```diff
@@ -36,7 +36,7 @@
 <template id="layout" name="Main layout" inherit_id="portal.frontend_layout">
     <xpath expr="//html" position="before">
-        <t t-set="editable_in_backend" t-value="edit_in_backend or ('website_published' in main_object.fields_get() and main_object._name != 'website.page')"/>
+        <t t-set="editable_in_backend" t-value="edit_in_backend"/>
```
Why not. Did someone asked for this change? Although, would the variable "editable_in_backend" still useful in that case?

Guillaume-gdi replied:
Indeed the variable "editable_in_backend" is useless now, I will remove it.

qsm-odoo replied:
> Did someone asked for this change?

Ok, now I know why you did not answer, that's just buggy without it... :disappointed: I'll restore it alongside blog post button which is not worth nor consistent to remove. Confirmed with PDE.

pierre-pde commented:
@Guillaume-gdi , looks great to me! thanks

qsm-odoo reviewed (changes requested):
Some remarks then I'll merge :+1: 

I tested the end result, I like it. I was a bit worried (less than @rdeodoo but still) that the flicker of the button appearing would be annoying but in the end, I think it's not a problem at all. We can always rethink/improve later (or after 16.0) but it's definitely mergeable with this design IMO :+1:

Guillaume-gdi commented on addons/website/static/src/systray_items/edit_in_backend.js:
```diff
@@ -35,16 +26,19 @@ export class EditInBackendSystray extends Component {
+    get label() {
+        const modelName = this.websiteService.currentWebsite.metadata.mainObject.model;
+        const objectName = this.websiteService.currentWebsite.metadata.mainObjectName;
+        return modelName === 'appointment.type' ? this.env._t('Appointment') :
+            modelName === 'product.template' ? this.env._t('Product') : objectName;
```
IMHO, 'Appointment Type' is also strange for the end user.. but I can live with that. For the rest I agree with you :)

pierre-pde replied:
@Williambraecky , can you think of a bad impact with the changes suggested by Quentin? 
Functional name change:
Product Template => Product
Product => Product Variant
I'm not sure where this is used.

qsm-odoo replied:
Asked in MP in YTI's channel: ok. As for appointment, "Appointment Type" is really not strange IMO. It's what we edit, not the appointment a visitor made.

qsm-odoo reviewed (approved):
@Guillaume-gdi Next time we debate for days about the translation of an element... maybe try to test your feature with a different language to see if what you did works? It did not work at all so I basically had to change everything. There were two problems:
- You used `_description` (while apparently @rdeodoo explained you not to) which does not retrieve the translated version.
- You added the model name inside the iframe. So even if you used the right stuff to get the translated version, it would be done with the website language instead of the backend UI language.

In the end, this raised problems as that model has to be retrieved from the JS and no suitable method is available to do that. I have to use a temporary stuff so we can move on and I'll have to check with the framework-py later to make something right (something that you could have done in time yourself within this PR directly if you tested your own work).

Anyways, I also reviewed the specs as it was not consistent: reaching an appointment type page only showed the button on the calendar page instead of all pages, confusing. I restored the blog post button too as it will be the only button to reach that form later on. I also squashed both commit as it does not make sense to fix something that you remove in the commit just after. I kept your explanation in the squashed commit message. Here is the diff I made: https://github.com/odoo/odoo/compare/a98d15b35a058f7344a2d6900e87203e1c02bf3c..5982d667652a76665d134a8a904f00fc89959b3b

In the end, there is also another bug but I could swear it was not there friday so maybe it is very new (and it is in master, not because of this PR): the edit in backend button does not appear for the demo user once you add it manager right on something (I tested with Events). There is no data-main-object or anything on the HTML :shrug: cc @rdeodoo I would not be surprised if this is a t-cache issue :shrug: 

We need to move on, I noted everything that have to be checked / fixed / changed but let's already merge this so people can start making videos of this UI :shrug: 

@robodoo r+

younn-o commented on addons/website/static/src/systray_items/edit_in_backend.js:
```diff
@@ -1,29 +1,25 @@
         this.websiteService = useService('website');
         this.actionService = useService('action');
 
+        onWillStart(this._updateMainObjectName);
+        useBus(websiteSystrayRegistry, 'CONTENT-UPDATED', this._updateMainObjectName);
```
Hello, the onWillStart will not be called again when switching between two pages which have a main object, because the component is not unmounted when doing that (and it has no props or state so it is not supposed to rerender).
The `CONTENT-UPDATED` event is sent when there is a new pageDocument so that the navbar rerenders. When the previous and the current pageDocument have both the editableInBackend metadata, the systray item will remain as displayed in the website systray registry, and onWillUnmount and onWillStart will not be called as nothing changed.
But when the new pageDocument has not that metadata, the navbar will rerender without that systray item and it will be unmounted, so on the next pageDocument with editableInBackend it will we mounted again and onWillStart will be called.

I think it would be better to define the mainObjectName variable as a state variable, as we want to rerender the component when it changes (I have to admit I don't understand why the component is rerendered with the correct mainObjectName right now, the `CONTENT-UPDATED` should only update the variable, so something else triggers a rerender of the component, but I don't think it is "intentional", as the component has no state nor props, so it may change if it's coming from a bug and it gets fixed. It is visible with a slower `getUserModelName` method).

qsm-odoo replied:
> I have to admit I don't understand why the component is rerendered with the correct mainObjectName right now

Yes, that's what I did not understand. But I guess we do not have to understand and just use a state variable indeed. @Guillaume-gdi would you handle that?

Guillaume-gdi replied:
Yes I'll handle it :+1:

Guillaume-gdi replied:
@qsm-odoo Here it is: https://github.com/odoo/odoo/pull/99599
